LDAP RFC4519 group membership

adamruzicka opened the pull request:
Companion to https://github.com/theforeman/foreman/pull/10527 . Draft until the main PR settles

adamruzicka committed:
LDAP RFC4519 group membership

evgeni commented on plugins/modules/auth_source_ldap.py:
```diff
@@ -215,14 +215,19 @@ def main():
     # additional parameter checks
-    if 'use_netgroups' in module.foreman_params and module.foreman_params['server_type'] == 'active_directory':
-        module.fail_json(msg='use_netgroups cannot be used when server_type=active_directory')
+    server_type = module.foreman_params['server_type']
```
server_type is not a required param, so this can be absent.
```suggestion
    server_type = module.foreman_params.get('server_type', 'posix')
```

adamruzicka replied:
Oh, right, fixed

evgeni commented on plugins/modules/auth_source_ldap.py:
```diff
@@ -215,14 +215,19 @@ def main():
             server_type=dict(choices=["free_ipa", "active_directory", "posix"]),
             ldap_filter=dict(),
-            use_netgroups=dict(type='bool'),
+            ldap_group_membership=dict(choices=["posix", "nis_netgroups", "rfc4519"]),
```
In a similar spirit as https://github.com/theforeman/foreman/pull/10527#discussion_r2163538370, if we try to be backwards compatible, we could allow both to be provided (but exclusively), and translate depending on the server accepting the new thing or not.

adamruzicka replied:
Copied most of the logic from the main foreman pr

adamruzicka replied:
> Copied most of the logic from the main foreman pr

Which wasn't really the greatest of ideas, changed again.

adamruzicka committed:
use_netgroups and ldap_group_membership compatibility

ldap_group_membership (LGM) wins over use_netgroups, if both are passed
in. If LGM is unsupported by the server, user-provided use_netgroups is
used. If LGM is unsupported by the server, but no use_netgroups is
provided by the user, use_netgroups is derived from LGM.

adamruzicka commented:
🟢

evgeni commented on plugins/modules/auth_source_ldap.py:
```diff
@@ -102,9 +102,16 @@
-    description: Whether to use NIS netgroups instead of posix groups, not valid for I(server_type=active_directory)
+    description:
+      - Whether to use NIS netgroups instead of posix groups, not valid for I(server_type=active_directory)
+      - "Deprecated: The use_netgroups parameter is deprecated with Foreman 3.16 in favor of I(ldap_group_membership) and will be removed in a future release."
```
```suggestion
      - "Deprecated: The I(use_netgroups) parameter is deprecated with Foreman 3.16 in favor of I(ldap_group_membership) and will be removed in a future release."
```

evgeni reviewed (approved):
Ack with a styling nitpick :)

evgeni commented:
argh, line too long what

adamruzicka commented:
Whoops, might have accidentally overwritten your changes
